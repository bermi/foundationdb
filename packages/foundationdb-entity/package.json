{
    "name": "@openland/foundationdb-entity",
    "description": "Entity layer for FoundationDB",
    "version": "11.5.3",
    "main": "lib/index.js",
    "typings": "lib/index.d.ts",
    "repository": "https://github.com/openland/foundationdb",
    "author": "Steve Kite <steve@openland.com>",
    "license": "MIT",
    "gitHead": "8ec0134f55301034ec512aedb7c7c30b9a993235",
    "publishConfig": {
        "access": "public"
    },
    "dependencies": {
        "@openland/context": "*",
        "@openland/foundationdb": "^11.5.3",
        "@openland/foundationdb-bus": "^11.5.3",
        "@openland/foundationdb-utils": "^11.5.3",
        "@openland/lifetime": "*",
        "@openland/log": "*"
    }
}
