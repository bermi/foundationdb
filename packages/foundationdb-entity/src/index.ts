export { AtomicBoolean } from './AtomicBoolean';
export { AtomicBooleanFactory } from './AtomicBooleanFactory';
export { AtomicInteger } from './AtomicInteger';
export { AtomicIntegerFactory } from './AtomicIntegerFactory';
export { EntityStorage } from './EntityStorage';
export { BaseStore } from './BaseStore';
export { Entity } from './Entity';
export { EntityFactory } from './EntityFactory';
export { EntityMetadata } from './EntityMetadata';
export { EntityDescriptor, SecondaryIndexDescriptor, PrimaryKeyDescriptor } from './EntityDescriptor';
export { codecs, Codec, typeSymbol } from './codecs';
export { ShapeWithMetadata } from './ShapeWithMetadata';
export { PrimaryKeyType } from './PrimaryKeyType';