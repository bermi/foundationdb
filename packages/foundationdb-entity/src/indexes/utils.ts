import { IndexField } from '../EntityDescriptor';
import * as Tuple from '@openland/foundationdb-tuple';
import { encoders, TupleItem } from '@openland/foundationdb';

export function tupleToCursor(tuple: TupleItem[]) {
    return encoders.tuple.pack(tuple).toString('hex');
}

export function cursorToTuple(src: string) {
    return encoders.tuple.unpack(Buffer.from(src, 'hex'));
}

export function tupleKey(tuple: TupleItem[]) {
    return encoders.tuple.pack(tuple).toString('hex');
}

export function resolveIndexKey(src: any[], fields: IndexField[], partial: boolean = false, offset: number = 0) {
    if (!partial && src.length !== fields.length) {
        throw Error('Invalid key length');
    }

    let res: Tuple.TupleItem[] = [];
    for (let i = 0; i < src.length; i++) {
        let key = fields[i + offset];
        let v = src[i];
        if (key.type === 'opt_boolean'
            || key.type === 'opt_string'
            || key.type === 'opt_float'
            || key.type === 'opt_integer'
        ) {
            if (v === null) {
                res.push(v);
                continue;
            }
        }
        if (key.type === 'boolean' || key.type === 'opt_boolean') {
            if (typeof v !== 'boolean') {
                throw Error('Unexpected key');
            }
            res.push(v);
        } else if (key.type === 'integer' || key.type === 'opt_integer') {
            if (typeof v !== 'number') {
                throw Error('Unexpected key');
            }
            res.push(v);
        } else if (key.type === 'float' || key.type === 'opt_float') {
            if (typeof v !== 'number') {
                throw Error('Unexpected key');
            }
            res.push(new Tuple.Float(v as number));
        } else if (key.type === 'string' || key.type === 'opt_string') {
            if (typeof v !== 'string') {
                throw Error('Unexpected key');
            }
            res.push(v);
        } else {
            throw Error('Unknown index key');
        }
    }
    return res;
}