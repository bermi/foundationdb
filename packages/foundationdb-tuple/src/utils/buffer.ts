
export class BufferReader {
    readonly bufffer: Buffer;
    offset: number = 0;
    constructor(src: Buffer) {
        this.bufffer = src;
    }

    get completed() {
        return this.offset >= this.bufffer.length;
    }

    peek() {
        if (this.completed) {
            throw Error('EOF');
        }
        return this.bufffer.readUInt8(this.offset);
    }

    readByte() {
        if (this.completed) {
            throw Error('EOF');
        }
        return this.bufffer.readUInt8(this.offset++);
    }

    readBuffer(size: number) {
        if (this.completed) {
            throw Error('EOF');
        }
        if (this.offset + size > this.bufffer.length) {
            throw Error('EOF');
        }
        let res = this.bufffer.slice(this.offset, this.offset + size);
        this.skip(size);
        return res;
    }

    expect(byte: number) {
        let res = this.bufffer.readUInt8(this.offset++);
        if (res !== byte) {
            throw Error('Expected ' + byte + ', got ' + res);
        }
    }
    skip(count: number) {
        this.offset += count;
    }
}

export class BufferWriter {
    private _storage: Buffer;
    private _used: number = 0;

    constructor(capacity: number = 64) {
        this._storage = Buffer.alloc(capacity);
    }

    writeByte(val: number) {
        this._need(1); this._storage[this._used++] = val;
    }

    writeString(val: string) {
        const len = Buffer.byteLength(val);
        this._need(len);
        this._storage.write(val, this._used);
        this._used += len;
    }

    writeBuffer(val: Buffer) {
        this._need(val.length);
        val.copy(this._storage, this._used);
        this._used += val.length;
    }

    build() {
        const result = Buffer.alloc(this._used);
        this._storage.copy(result, 0, 0, this._used);
        return result;
    }

    private _need(numBytes: number) {
        if (this._storage.length < this._used + numBytes) {
            let newAmt = this._storage.length;
            while (newAmt < this._used + numBytes) {
                newAmt *= 2;
            }
            const newStorage = Buffer.alloc(newAmt);
            this._storage.copy(newStorage);
            this._storage = newStorage;
        }
    }
}