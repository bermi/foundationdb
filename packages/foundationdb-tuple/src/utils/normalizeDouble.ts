export function normalizeDouble(src: number) {
    if (!Number.isFinite(src)) {
        throw Error('Number ' + src + ' is not finite');
    }
    return src;
}