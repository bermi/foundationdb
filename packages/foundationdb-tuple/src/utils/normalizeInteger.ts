/**
 * This function is intended to detect if number is integer that 
 * can be safely encoded as integer
 * 
 * @param src sorce number
 */
export function normalizeInteger(src: number): number {
    if (typeof src !== 'number') {
        throw Error('Input is not integer');
    }
    if (!Number.isFinite(src)) {
        throw Error('Number ' + src + ' is not finite');
    }
    if (!Number.isInteger(src)) {
        throw Error('Number ' + src + ' is not an integer');
    }
    if (!Number.isSafeInteger(src)) {
        throw Error('Number ' + src + ' is not a safe integer');
    }
    return src | 0;
}