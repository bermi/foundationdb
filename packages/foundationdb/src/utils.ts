const byteZero = Buffer.alloc(1)
byteZero.writeUInt8(0, 0)

/**
 * Getting an "increment" of a key. 
 * 
 * Increment of a key is a key that is ordered after all keys that is prefixed by a first one.
 * For example, keyIncrement(ABCD) -> ABCE.
 * 
 * @param buf source key
 */
export function keyIncrement(buf: Buffer): Buffer {
    let lastNonFFByte
    for (lastNonFFByte = buf.length - 1; lastNonFFByte >= 0; --lastNonFFByte) {
        if (buf[lastNonFFByte] !== 0xFF) {
            break;
        }
    }

    if (lastNonFFByte < 0) {
        throw new Error(`invalid argument '${buf}': prefix must have at least one byte not equal to 0xFF`)
    }

    const result = Buffer.alloc(lastNonFFByte + 1)
    buf.copy(result, 0, 0, result.length)
    ++result[lastNonFFByte]

    return result;
}

/**
 * Calculating next key. Next key is a first key that right after provided one. It is always `key + 0x00`.
 * @param buf source key
 */
export function keyNext(buf: Buffer): Buffer {
    // Buffer.from does support taking a string but @types/node has overly
    // strict type definitions for the function.
    return Buffer.concat([buf, byteZero], buf.length + 1)
}