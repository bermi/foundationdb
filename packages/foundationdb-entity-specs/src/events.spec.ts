import { EntityStorage } from '@openland/foundationdb-entity';
import { openTestDatabase } from './utils/openTestDatabase';
import { openStore, SampleEvent } from './events.repo';

describe('Events', () => {
    it('should decode and encode events', async () => {
        let db = await openTestDatabase();
        let store = await openStore(new EntityStorage(db));
        let encoded = store.eventFactory.encode(SampleEvent.create({
            id: 'some_id',
            name: 'name'
        }));
        expect(encoded).toMatchSnapshot();
        let decoded = store.eventFactory.decode(encoded);
        expect(decoded).toMatchSnapshot();
    });
});