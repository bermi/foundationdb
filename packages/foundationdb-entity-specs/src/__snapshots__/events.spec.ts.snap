// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Events should decode and encode events 1`] = `
Object {
  "data": Object {
    "id": "some_id",
    "name": "name",
  },
  "type": "sampleEvent",
}
`;

exports[`Events should decode and encode events 2`] = `
SampleEvent {
  "raw": Object {
    "id": "some_id",
    "name": "name",
  },
  "type": "sampleEvent",
}
`;
